channeldb/payment_control.go

```diff
@@ -615,19 +619,45 @@ func fetchPaymentBucketUpdate(tx kvdb.RwTx, paymentHash lntypes.Hash) (
 
 // nextPaymentSequence returns the next sequence number to store for a new
 // payment.
-func nextPaymentSequence(tx kvdb.RwTx) ([]byte, error) {
-	payments, err := tx.CreateTopLevelBucket(paymentsRootBucket)
-	if err != nil {
-		return nil, err
-	}
+func (p *PaymentControl) nextPaymentSequence() ([]byte, error) {
+	p.paymentSeqMx.Lock()
+	defer p.paymentSeqMx.Unlock()
+
+	// Set a new upper bound in the DB every 1000 payments to avoid
+	// conflicts on the sequence when using etcd.
+	if p.currPaymentSeq == p.storedPaymentSeq {
+		var currPaymentSeq, newUpperBound uint64
+		if err := kvdb.Update(p.db.Backend, func(tx kvdb.RwTx) error {
+			paymentsBucket, err := tx.CreateTopLevelBucket(
+				paymentsRootBucket,
+			)
+			if err != nil {
+				return err
+			}
 
-	seq, err := payments.NextSequence()
-	if err != nil {
-		return nil, err
+			currPaymentSeq = paymentsBucket.Sequence()
```

Comment:
So we might end up skipping some of the integers right? Does the consistency in the sequence matter? If not, we might as well use `UUID` to skip the db query in the future?

Imagine a very very edge case that,
- a new `lnd` is setup, meaning we have `paymentsBucket.Sequence()` being 0.
- our `currPaymentSeq` is 0, and `newUpperBound` is 1000.
- then the user restarts `lnd`, again lands here.
- now our `currPaymentSeq` is 1000, and we've skipped the sequence 0 to 1000.

A more likely case is we will constantly skip a total number of integers from 0 to 1000.

Reply:
Yes I agree using UUIDs were probably better but that implies migration as well as RPC changes. I think we can live with the gaps although perhaps we should need to rename the sequence to id or something to avoid implying increment by one vs being monotonically increasing.